### DISCUSSION TITLE
Doubts about get_inspiral_tf

### DISCUSSION THREAD

**Discussion by emmeDG** [gwastro/pycbc#4677] [Q&A]
Hi, 
I have a doubt about the get_inspiral_tf function.
In theory, this should output the TF evolution of a signal given the approximant, but it is not clear to me what it really does because I get some inconsistencies.

For example, I generate the TF evolution of a signal in this way

`time_tf, tf_evo = get_inspiral_tf(time, 30, 30, spin1 = 0, spin2 = 0, f_low = 20, n_points=2000, pn_2order=7, approximant='TaylorF2')`

And get the following plot
![plot](https://github.com/gwastro/pycbc/assets/72254807/fec95ecc-a775-4751-867d-4909a586efbf)

For what I understand, this is a signal that lasts about 0.8 seconds with a frequency sweep up to 70 Hz.

 Then, I generate a TD signal and plot it

```
hp, hc = get_td_waveform(approximant='TaylorF2',
                         mass1=30,
                         mass2=30,
                         inclination=0,
                         #coa_phase=2.45,
                         delta_t=1.0/(100),
                         f_lower=20,
                         distance = 40)

hp.plot()

```
I get this:

![plot2](https://github.com/gwastro/pycbc/assets/72254807/17f3b77e-16a8-4a8a-a35d-78720c5bad73)

 That is a waveform that, up to merger lasts almost 5 seconds. Which is the difference between the two outputs?.

Please note that, if I transform the latter into a frequency series and/or generate a wf in FD, I correctly get frequencies up to 70 Hz.

**Comment by spxiwh** [comment]
I think that what you are seeing here in the second plot is Gibbs' noise (https://en.wikipedia.org/wiki/Gibbs_phenomenon). Your plot looks a lot worse than what I get when running this:

![image](https://github.com/gwastro/pycbc/assets/11438929/c97a0f6d-1a50-4c61-a0e7-70c8d5abaf05)

but the issues are still present here. In particular:

* TaylorF2 waveforms do not have any form of merger/ringdown, they just stop, abruptly at some frequency ... and start, abruptly, at some frequency. This can cause some nasty effects in the time-domain, which aren't *necessarily* bad, but you do have to be careful with this. The biggest issue here would be that the waveform is stopping abruptly at O(100Hz) where a lot of power would be in reality. A waveform including merger+ringdown effects would be better. Compare (for example) to using SEOBNRv4 as the approximant:

![image](https://github.com/gwastro/pycbc/assets/11438929/d8aea11c-ebda-4f7a-abc0-027286aa682d)

(this is a time-domain waveform, so it starts abruptly in the time, not the frequency, domain). Using PhenomD:

![image](https://github.com/gwastro/pycbc/assets/11438929/5eb7f0f6-7fcf-4c74-83e7-441e30f06f1b)

Still has some Gibbs noise from the abrupt start of that waveform *in the frequency domain*.

* The sample rate used here is too small. With a delta_t of 0.01s, the Nyquist frequency is 50Hz, and these waveforms are extending past that. SEOB will fail, because it knows this is a problem, but TaylorF2 does not fail, and I'm not really clear how it removes content above 50Hz (if it is aliased, or if a proper low-pass is applied). In either case, make it larger. I used 0.001s to make the plots above.

**Reply by emmeDG** [reply]
Hi Ian,
thank you for your reply.

In fact, for the analysis that I want to make, I am interested in the low frequency part of the waveform only. Let's say in the range 2-10 Hz. Which is the best approximant to use there? Is there also a table or a summary somewhere with all the features of each approximant?

In general, for the real analysis, I would also use higher sample rates. The problem is, since I am interested in generating a waveform that starts at 2 Hz, the generation of the waveform takes a while and for purely testing purposes of the code I would like to keep the things faster.

**Reply by spxiwh** [reply]
There likely won't be a lot of difference between waveforms at a total mass of 60 solar masses and a frequency range of 2 - 10 Hz. However PyCBC uses `lalsimulation` directly to generate most waveforms. The documentation for `lalsimulation` is here https://lscsoft.docs.ligo.org/lalsuite/lalsimulation/, but I'm not sure it has what you want! There is a list of available waveforms here https://lscsoft.docs.ligo.org/lalsuite/lalsimulation/group___l_a_l_sim_inspiral__h.html#gab955e4603c588fe19b39e47870a7b69c but it doesn't distinguish between "this is old, never use it" and "this is good", nor does it discuss much about the details. The question of "what is best" really depends on what you're wanting to do (ie. do you want precession, higher order modes, eccentricity?), but perhaps `IMRPhenomXAS` is a good choice (if you don't want to consider precession/higher order modes or eccentricity).

For a frequency-domain waveform like `IMRPhenomXAS` or `TaylorF2` you can generate directly in the frequency domain (using `get_fd_waveform`), this will be quicker, especially if whatever comes next is done also in the frequency domain.

**Reply by emmeDG** [reply]
Yes, that is exactly what I needed!
I think that now everything is more clear.

I have just one very last question: since generating a low mass (less than 2 m_sun) from low frequency (2 Hz) TD waveform can take a while, do you think it is appropriate to generate the FD waveform, select the frequency sub-band that I am interested in and then transform it into the TD using the ifft method embedded into the frequency series? Or would I get inappropriate numerical errors?

Thank you!
